src/java.base/share/classes/java/io/Reader.java

```diff
@@ -457,7 +448,64 @@ private String readAllCharsAsString() throws IOException {
      * @since 25
      */
     public List<String> readAllLines() throws IOException {
-        return readAllCharsAsString().lines().toList();
+        char[] cb = new char[TRANSFER_BUFFER_SIZE];
+        int pos = 0;
+        List<String> lines = new ArrayList<String>();
+
+        StringBuilder sb = new StringBuilder(82);
+        int n = read(cb, 0, cb.length);
+        boolean eos = (n == -1);
+        while (!eos) {
+            boolean eol = false;
+            boolean stringAdded = false;
+            while (!eol) {
+                // Find the next line terminator. If none is found,
+                // "term" will equal "n".
+                int term = pos;
+                while (term < n) {
+                    char c = cb[term];
+                    if (c == '\r' || c == '\n')
+                        break;
+                    term++;
+                }
+
+                // Terminator found so at EOL.
+                if (term < n)
+                    eol = true;
+
+                if (term == pos) {
+                    // Current position is terminator so skip it.
+                    pos++;
+                } else { // term > pos
+                    if (eol && sb.length() == 0) {
+                        // Create and add a string to avoid the StringBuilder.
+                        lines.add(new String(cb, pos, term - pos));
+                        stringAdded = true;
+                    } else
+                        sb.append(cb, pos, term - pos);
+                    pos = term + 1;
+                }
+
+                if (pos >= n) {
+                    // Buffer content consumed so reload it.
+                    if ((n = read(cb, 0, cb.length)) < 0) {
+                        eos = eol = true;
+                        break;
+                    }
+                    pos = 0;
+                }
+            }
+
+            if (!stringAdded) {
```

Comment:
Can we change the maintenancce of `stringAdded` condition to a check `sb.length() == 0`?

Reply:
> Edit: No, we need to check empty lines.

Right. I caught this during testing.